### PR TITLE
add blockwise rotation.

### PR DESCRIPTION
**Pull request by lkk12014402**
## Description

This change adds **block-wise (layer-wise) rotation** to AutoRound's QuaRot/SpinQuant
rotation quantization, enabling rotation to be applied **per decoder block, in lock-step
with AutoRound's block-wise quantization**, instead of rotating the whole model up front.

Enabled via a single top-level switch:

```python
AutoRound(model, scheme="MXFP4", rotation_config="quarot", layerwise_rotation=True)
```

## Motivation

The standard QuaRot/SpinQuant flow rotates the **entire model** first (R1 fused into
embedding/all linears, R2 into attention, R3/R4 as online hooks), then quantizes. This:

- requires the whole model to be resident in memory at once (heavy for large models), and
- is decoupled from AutoRound's block-wise quantization pipeline.

Block-wise rotation aligns rotation with block-wise quantization so only one block needs
to be on-device at a time.

### CODE REVIEW

**Comment by lkk12014402** [comment]
## Accuracy results — Qwen3-8B · MXFP4 (RTN, iters=0) · PIQA

| # | Config | rotation_config | layerwise | acc | acc_norm |
|---|--------|-----------------|:---------:|-------|----------|
| 1 | MXFP4 baseline | — | — | 0.7312 | 0.7454 |
| 2 | Hadamard | `"hadamard"` | ✗ | 0.7416 | 0.7601 |
| 3 | random-Hadamard | `{"algorithm":"hadamard","hadamard_type":"random_hadamard"}` | ✗ | **0.7541** | **0.7753** |
| 4 | QuaRot R1R2 | `"quarot"` | ✗ | 0.7437 | 0.7508 |
| 5 | QuaRot R1R2R3R4 | `{"algorithm":"spinquant","r1..r4":true}` | ✗ | 0.7312 | 0.7394 |
| 6 | **block-wise QuaRot R1R2** | `"quarot"` | ✓ | 0.7437 | 0.7508 |
| 7 | **block-wise QuaRot R1R2R3R4** | `{"algorithm":"spinquant","r1..r4":true}` | ✓ | 0.7312 | 0.7394 |

**Comment by wenhuach21** [comment]
Could you help move hadamard's entry (apply_rotation) to composer.py, I guess placing it in prepare_run  function is doable https://github.com/intel/auto-round/blob/main/auto_round/algorithms/composer.py#L542

File: auto_round/compressors/base.py [hunk 1]
```diff
@@ -1100,36 +1114,109 @@ def _resolve_formats(self) -> None:
...
                 self.model_context.model,
                 rotation_cfg,
                 data_type=self.quantize_config.data_type,
             )
 
+    # ------------------------------------------------------------------
+    # Block Lifecycle Hooks (layer-wise / block-wise rotation)
+    # ------------------------------------------------------------------
+
+    def _on_block_ready(self, block: torch.nn.Module, block_name, block_idx: int) -> None:
```
Comment (wenhuach21) [inline auto_round/compressors/base.py]: better  follow the new arch and move the rotation func to rotation folder

File: auto_round/compressors/orchestrator.py [hunk 2]
```diff
@@ -231,6 +231,9 @@ def _quantize_blocks(
 
             m, _, _ = self.alg_composer.dispatch_block(m, input_ids, input_others)
 
+            # ── Layer-wise rotation: rotate this block before reference collection ──
+            self._on_block_ready(m, block_name_or_names, i)
```
Comment (wenhuach21) [inline auto_round/compressors/orchestrator.py]: Please follow the architecture to decouple the different algorithms. You can discuss with Heng how to support this feature with less coupling.

File: auto_round/compressors/orchestrator.py [hunk 3]
```diff
@@ -433,6 +441,8 @@ def _quantize_zero_shot(self) -> tuple[torch.nn.Module, dict[str, Any]]:
                 memory_monitor.log_summary()
                 pbar.update(1)
 
+        self._finalize_block_processing(self.model)
```
Comment (wenhuach21) [inline auto_round/compressors/orchestrator.py]: similar issue

**Comment by wenhuach21** [comment]
Also, please fix this issue ASAP: https://github.com/intel/auto-round/issues/2036. It's a critical issue that should be prioritized.

**Comment by lkk12014402** [comment]
fix the coupling problem of rotation, please help review again @wenhuach21  @n1ck-guo

File: auto_round/algorithms/composer.py [hunk 4]
```diff
@@ -545,3 +563,110 @@ def prepare_run(self, composer: "AlgorithmComposer" = None):
...
+                normalised = normalize_rotation_config(rotation_cfg)
+                if normalised is None:
+                    continue
+                rotation = BaseRotation.from_config(normalised)
+                if rotation.supports_layerwise:
+                    logger.info(
+                        "[Rotation] Layer-wise mode: preparing R matrices only "
+                        "(rotation deferred to per-block hook)."
+                    )
+                    rotation.prepare_layerwise(model, data_type=data_type)
```
Comment (wenhuach21) [inline auto_round/algorithms/composer.py:630]: the most parts of this function should  be handled by  hadamard itself.
the perfect version should be:
def rotate_model():
for ratation in self._rotation_transforms:
rotate_model

File: auto_round/algorithms/composer.py [hunk 5]
```diff
@@ -545,3 +563,110 @@ def prepare_run(self, composer: "AlgorithmComposer" = None):
     def finalize_run(self):
         for alg in self.members():
             alg.finalize_run()
+        # Rotation teardown is part of the model-level finalize stage.
+        self._finalize_rotation(self._owning_model())
```
Comment (wenhuach21) [inline auto_round/algorithms/composer.py:567]: rotation should be a member of alg, and override the fininal_run

File: auto_round/algorithms/composer.py [hunk 6]
```diff
@@ -197,6 +205,10 @@ def __init__(self, configs: list, orchestrator: "BaseOrchestrator" = None) -> No
                 self.block_quantizer.bind_block_forward_runner(self.block_forward)
         self.scheme = getattr(orchestrator, "scheme_context", None)
 
+        # Rotation lifecycle state (populated by apply_model_transforms)
+        self._rotation_transforms: list = []
+        self._rotation_prepared: bool = False
```
Comment (wenhuach21) [inline auto_round/algorithms/composer.py:210]: if rotate_model implentation is controlled by hadamard it self, then we could delete this arg, right?

File: auto_round/algorithms/composer.py (hunk 4 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py:630]: ok, I see.

File: auto_round/algorithms/composer.py (hunk 6 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py:210]: yes

File: auto_round/compressors/orchestrator.py [hunk 7]
```diff
@@ -382,12 +384,15 @@ def _quantize_zero_shot(self) -> tuple[torch.nn.Module, dict[str, Any]]:
...
+                # ``block_index`` carries the global block index so compress_block
+                # can drive layer-wise rotation with the correct layer_idx.
                 ctx = BlockContext(
                     model=self.model,
                     block_names=[block_name],
                     block_name=block_name,
-                    block_index=0,
+                    block_index=_zs_block_idx,
                 )
+                _zs_block_idx += 1
```
Comment (wenhuach21) [inline auto_round/compressors/orchestrator.py:395]: please be care of we have an arg named `nblocks` which will fuse multi-blocks to 1

File: auto_round/algorithms/composer.py [hunk 8]
```diff
@@ -545,3 +563,110 @@ def prepare_run(self, composer: "AlgorithmComposer" = None):
...
+    # Rotation is a model-level pre-quantisation transform. Full-model rotation
+    # must run *before* calibration data is cached, which is earlier than the
+    # per-member ``prepare_run`` stage; layer-wise rotation instead prepares its
+    # matrices here and rotates each block from within ``compress_block``. Both
+    # are driven internally so the orchestrator only calls the single generic
+    # entry point :meth:`apply_model_transforms`.
+
+    def _resolve_rotation_data_type(self) -> str:
+        """Best-effort resolution of the quantization data_type for rotation dispatch."""
+        if self.scheme is not None and getattr(self.scheme, "data_type", None):
```
Comment (wenhuach21) [inline auto_round/algorithms/composer.py:582]: same, all theses should be moved to hadamard folder

**Comment by wenhuach21** [comment]
Thanks for the refinements. The version looks much better now. It would be better to move the major details into the Hadamard implementation itself.

File: auto_round/compressors/base.py (hunk 1 above)
Comment (lkk12014402) [reply auto_round/compressors/base.py]: `_on_block_ready` has been removed from `compressors/base.py` entirely. Per-block rotation now lives in the rotation folder: `RotationPreprocessor.on_block_ready` (`algorithms/transforms/member.py`) delegates to `HadamardRotation.rotate_layer` (`algorithms/transforms/hadamard/apply.py`). The compressor no longer knows anything about rotation.

File: auto_round/compressors/orchestrator.py (hunk 2 above)
Comment (lkk12014402) [reply auto_round/compressors/orchestrator.py]: The orchestrator no longer calls any rotation hook — `_quantize_blocks` is rotation-agnostic.

File: auto_round/compressors/orchestrator.py (hunk 3 above)
Comment (lkk12014402) [reply auto_round/compressors/orchestrator.py]: `_finalize_block_processing` is gone. Rotation teardown now rides the normal member lifecycle: the orchestrator only calls `alg_composer.finalize_run()`, which loops `members()` and reaches `RotationPreprocessor.finalize_run()` → `HadamardRotation.finalize_layerwise()`. No rotation-specific finalize path in the orchestrator anymore.

File: auto_round/algorithms/composer.py (hunk 4 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py]: the composer no longer contains any Hadamard-specific logic.

File: auto_round/algorithms/composer.py (hunk 5 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py]: Rotation is now a first-class algorithm member — `RotationPreprocessor(BasePreprocessor)` — that appears in `members()`

File: auto_round/algorithms/composer.py (hunk 6 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py:239]: Now that the full-model-vs-layer-wise decision lives inside the member (it reads `config.layerwise` via `is_layerwise`), the redundant `layerwise` argument is gone from `rotate_model` and the call site no longer passes it.

File: auto_round/compressors/orchestrator.py (hunk 7 above)
Comment (lkk12014402) [reply auto_round/compressors/orchestrator.py:473]: Handled. each decoder layer receives the correct global `layer_idx` even when `nblocks > 1` fuses several blocks into one.

File: auto_round/algorithms/composer.py (hunk 8 above)
Comment (lkk12014402) [reply auto_round/algorithms/composer.py:623]: All the algorithm-specific pieces (`supports_layerwise`, `prepare_layerwise`, `rotate_layer`, `finalize_layerwise`) now live in `hadamard/apply.py`.

**Comment by n1ck-guo** [comment]
please help update the desciption. I think rotation_config is remove from AutoRound

**Review by n1ck-guo — APPROVED** [review]
LGTM
